### PR TITLE
[LDR-R1] EOS-14001: Handle motr timeout error by restart in case of high error rate

### PR DESCRIPTION
**Pull request by dsurnin**
Signed-off-by: Dmitrii Surnin <dmitrii.surnin@seagate.com>

### CODE REVIEW

File: s3config-test.yaml [hunk 1]
```diff
@@ -67,6 +67,8 @@ S3_SERVER_CONFIG:                                     # Section for S3 Server
    S3_SERVER_ENABLE_OBJECT_LEAK_TRACKING: false         # Enables the object leak feature changes
    S3_REDIS_SERVER_ADDRESS: "127.0.0.1"                 # In case if redis is used for kvs contains redis server address
    S3_REDIS_SERVER_PORT: 6379                           # In case if redis is used for kvs contains redis server port
+   S3_SERVER_MOTR_ETIMEDOUT_MAX_THRESHOLD: 5            # Number of ETIMEDOUT errors per monitoring window before s3server restart
+   S3_SERVER_MOTR_ETIMEDOUT_WINDOW_SEC: 60              # Monitoring window for motr ETIMEDOUT errors
```
Comment (knrajnambiar76) [inline s3config-test.yaml:71]: ...motr ETIMEDOUT errors in seconds

File: server/s3_motr_rw_common.cc [hunk 2]
```diff
@@ -89,7 +98,32 @@ void motr_op_done_on_main_thread(evutil_socket_t, short events,
...
+        gs_timeout_cnt = 0;
+      }
+      gs_timeout_cnt++;
+      if (gs_timeout_cnt >= err_thr) {
+        s3_log(
+            S3_LOG_ERROR, request_id,
+            "Shutdown. Motr ETIMEDOUT error cnt %u reached threshold value %u "
+            "for %" PRId64 " sec with allowed window %u sec",
+            gs_timeout_cnt, err_thr, curtime - gs_timeout_window_start,
+            err_wnd);
```
Comment (knrajnambiar76) [inline server/s3_motr_rw_common.cc:121]: you need to ensure that for some other request we dont call s3_kickoff_graceful_shutdown() again, so either reset gs_timeout_cnt and set err_thr to a very high value here or by some other way

File: server/s3_motr_rw_common.cc [hunk 3]
```diff
@@ -89,7 +98,32 @@ void motr_op_done_on_main_thread(evutil_socket_t, short events,
              S3_IEM_MOTR_CONN_FAIL_JSON);
     }
     context->on_failed_handler()();  // Invoke the handler.
+
+    if (error_code == -ETIMEDOUT) {
```
Comment (knrajnambiar76) [inline server/s3_motr_rw_common.cc:102]: Have a comment which talks about the logic

File: server/s3_option.cc [hunk 4]
```diff
@@ -1213,6 +1235,14 @@ std::string S3Option::get_redis_srv_addr() { return redis_srv_addr; }
 
 unsigned short S3Option::get_redis_srv_port() { return redis_srv_port; }
 
+unsigned S3Option::get_motr_etimedout_max_threshold() {
+  return motr_etimedout_max_threshold;
+};
```
Comment (knrajnambiar76) [inline server/s3_option.cc:1240]: remove those semicolon at end

File: s3config-test.yaml (hunk 1 above)
Comment (dsurnin) [reply s3config-test.yaml:71]: done

File: server/s3_motr_rw_common.cc (hunk 2 above)
Comment (dsurnin) [reply server/s3_motr_rw_common.cc:121]: done

File: server/s3_motr_rw_common.cc (hunk 3 above)
Comment (dsurnin) [reply server/s3_motr_rw_common.cc:102]: done

File: server/s3_option.cc (hunk 4 above)
Comment (dsurnin) [reply server/s3_option.cc:1240]: done

File: server/s3_common_utilities.cc [hunk 5]
```diff
@@ -179,3 +182,29 @@ std::string evhtp_error_flags_description(uint8_t errtype) {
 }
 
 }  // namespace S3CommonUtilities
+
+void s3_kickoff_graceful_shutdown(int ignore) {
+  extern int global_shutdown_in_progress;
```
Comment (amitwac2608) [inline server/s3_common_utilities.cc:187]: If time permits, we should make this variable as bool.. We are in C++ :)

File: server/s3_common_utilities.cc [hunk 6]
```diff
@@ -179,3 +182,29 @@ std::string evhtp_error_flags_description(uint8_t errtype) {
...
+    global_shutdown_in_progress = 1;
+    // signal handler
+    S3Option *option_instance = S3Option::get_instance();
+    int grace_period_sec = option_instance->get_s3_grace_period_sec();
+    struct timeval loopexit_timeout = {.tv_sec = 0, .tv_usec = 0};
+
+    // trigger rollbacks & stop handling new requests
+    option_instance->set_is_s3_shutting_down(true);
+
+    if (grace_period_sec > 6) {
```
Comment (amitwac2608) [inline server/s3_common_utilities.cc:200]: what is this "6" can we make it constant variable ?

File: server/s3_motr_rw_common.cc [hunk 7]
```diff
@@ -89,7 +101,39 @@ void motr_op_done_on_main_thread(evutil_socket_t, short events,
...
+      gs_timeout_window_start = curtime;
+      if (gs_timeout_cnt >= err_thr) {
+        s3_log(
+            S3_LOG_ERROR, request_id,
+            "Shutdown. Motr ETIMEDOUT error cnt %u reached threshold value %u "
+            "for %" PRId64 " sec with allowed window %u sec",
+            gs_timeout_cnt, err_thr, curtime - gs_timeout_window_start,
+            err_wnd);
+        s3_kickoff_graceful_shutdown(0);
+        gs_timeout_shutdown_in_progress = true;
```
Comment (amitwac2608) [inline server/s3_motr_rw_common.cc:132]: should we set this value before calling graceful_shutdown() ?

**Review by amitwac2608 — COMMENTED** [review]
Some minor comments/questions. Will be waiting for your response.

**Review by amitwac2608 — APPROVED** [review]

**Review by knrajnambiar76 — APPROVED** [review]

File: server/s3_common_utilities.cc (hunk 6 above)
Comment (dsurnin) [reply server/s3_common_utilities.cc]: It was deleted

File: server/s3_motr_rw_common.cc (hunk 7 above)
Comment (dsurnin) [reply server/s3_motr_rw_common.cc:136]: This flag indicates the function was called, so it is set after the call

File: server/s3_common_utilities.cc (hunk 5 above)
Comment (dsurnin) [reply server/s3_common_utilities.cc:187]: need to check with func author, will take some time

**Review by t7ko-seagate — APPROVED** [review]

**Pull request merged**